webui/react/src/stores/auth.tsx

```diff
@@ -0,0 +1,121 @@
+import React, { createContext, PropsWithChildren, useCallback, useContext, useState } from 'react';
+
+import { globalStorage } from 'globalStorage';
+import { Auth, DetailedUser } from 'types';
+import { getCookie, setCookie } from 'utils/browser';
+import { Loadable, Loaded, NotLoaded } from 'utils/loadable';
+
+export type CurrentUser = Auth & { checked: boolean };
+
+type AuthContext = {
+  auth: Loadable<CurrentUser>;
+  updateAuth: (fn: (users: Loadable<CurrentUser>) => Loadable<CurrentUser>) => void;
+};
+
+type UseAuthReturn = {
+  auth: Loadable<CurrentUser>;
+  resetAuth: () => void;
+  setAuth: (auth: Auth) => void;
+  setAuthCheck: () => void;
+  updateCurrentUser: (user: DetailedUser, users: DetailedUser[]) => void;
+};
+
+export const AUTH_COOKIE_KEY = 'auth';
+
+const AuthContext = createContext<AuthContext | null>(null);
+
+export const AuthProvider: React.FC<PropsWithChildren> = ({ children }) => {
+  const [auth, setAuth] = useState<Loadable<CurrentUser>>(NotLoaded);
+
+  return (
+    <AuthContext.Provider
+      value={{
+        auth,
+        updateAuth: setAuth,
+      }}>
+      {children}
+    </AuthContext.Provider>
+  );
+};
+
+const clearAuthCookie = (): void => {
+  document.cookie = `${AUTH_COOKIE_KEY}=; expires=Thu, 01 Jan 1970 00:00:00 UTC; path=/;`;
+};
+
+/**
+ * set the auth cookie if it's not already set.
+ * @param token auth token
+ */
+const ensureAuthCookieSet = (token: string): void => {
+  if (!getCookie(AUTH_COOKIE_KEY)) setCookie(AUTH_COOKIE_KEY, token);
+};
+
+export const useAuth = (): UseAuthReturn => {
+  const context = useContext(AuthContext);
+
+  if (context === null) {
+    throw new Error('Attempted to use useAuth outside of Auth Context');
+  }
+  const { auth, updateAuth } = context;
+
+  const updateCurrentUser = useCallback(
+    (user: DetailedUser, users: DetailedUser[]) => {
+      updateAuth((prevState) => {
+        if (!Loadable.isLoaded(prevState)) return prevState;
```

Comment:
This makes me think that `currentUser` should just be it's own bit of state since it is apparently updated independently of the rest of this object.